### PR TITLE
Use AzureClusterIdentity when running ci e2e tests

### PR DESCRIPTION
**Pull request by nader-ziada**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
- use AzureClusterIdentity in ci to make sure it working in all cases

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**TODOs**:

- [ ] squashed commits
- [ ] includes documentation
- [ ] adds unit tests

**Release note**:

```release-note
- Update aad-pod-identity to v1.8.0 which upgrades CRDs from apiextensions/v1beta1 to apiextensions/v1
- Update calico version used by templates and tests
```

### CODE REVIEW

**Comment by nader-ziada** [comment]
> the `private cluster` test specifically is failing every time, looking into it

I believe the private cluster test is failing because of this issue here https://azure.github.io/aad-pod-identity/docs/troubleshooting/#token-requests-calls-fail-with-io-timeout

**Comment by nader-ziada** [comment]
more info about issues with tests failing in this PR can be found in this slack thread https://kubernetes.slack.com/archives/CEX9HENG7/p1621362051002500

tldr; 
 - the nmi pod is having dns issues and can't authenticate when the management cluster is running on the Azure (not on a kind cluster) as in the case of the `private cluster` test. The dns issues happen when using `dnsPolicy=ClusterFirstWithHostNet` as recommended for any pod using `hostNetwork=true` 
 - The worker node has no issues connecting and the `nmi` pod works fine with `dnsPolicy=ClusterFirst`
 - but there seems to be an issue with either `how we configure the nmi pod`, or `the azure cluster vnet`

**Comment by nader-ziada** [comment]
So the real issue here is kernel bug we get stuck on when using calico with vxlan

more details here: https://github.com/projectcalico/calico/issues/3145

**Comment by nader-ziada** [comment]
I think we need to wait for the next release of calico that should include this fix here

https://github.com/projectcalico/felix/pull/2811

**Comment by nader-ziada** [comment]
removed the identity ref from the private cluster template for now

**Comment by nader-ziada** [comment]
@CecileRobertMichon ready for review

File: test/e2e/common.go [hunk 1]
```diff
@@ -51,6 +51,8 @@ const (
 	AzureCPSubnetCidr        = "AZURE_CP_SUBNET_CIDR"
 	AzureNodeSubnetCidr      = "AZURE_NODE_SUBNET_CIDR"
 	MultiTenancyIdentityName = "MULTI_TENANCY_IDENTITY_NAME"
+	ClusterIdentityName      = "CLUSTER_IDENTITY_NAME"
+	VMSSHPort                = "VM_SSH_PORT"
```
Comment (CecileRobertMichon) [inline test/e2e/common.go:55]: are you using VMSSHPort anywhere? It was removed in another PR

Comment (nader-ziada) [reply test/e2e/common.go:55]: looks like the rebase messed up something, will clean up

Comment (nader-ziada) [reply test/e2e/common.go]: done

**Review by CecileRobertMichon — COMMENTED** [review]
should we get rid of the `different SP identity` test since it's now equivalent to the others?

**Comment by nader-ziada** [comment]
> should we get rid of the `different SP identity` test since it's now equivalent to the others?

yes, will remove that

**Comment by fiunchinho** [comment]
I see that the release note for this PR is `None` but it's upgrading `aad-pod-identity` to `v1.8.0` which upgrades CRDs from apiextensions/v1beta1 to apiextensions/v1. Should this be included in the release note?

**Comment by nader-ziada** [comment]
> I see that the release note for this PR is `None` but it's upgrading `aad-pod-identity` to `v1.8.0` which upgrades CRDs from apiextensions/v1beta1 to apiextensions/v1. Should this be included in the release note?

good point, it really should have been a different PR, but was done as part fixing this up here, will update release notes for now and if reviewers feel I should split it out in a different PR, I can do that

**Review by CecileRobertMichon — COMMENTED** [review]
/lgtm
/assign @devigned 

@nader-ziada do you have any thoughts on how we can encourage users to use AzureClusterIdentites for their v1alpha4 clusters? Should the default template example have it? Related to #1445

**Comment by nader-ziada** [comment]
> @nader-ziada do you have any thoughts on how we can encourage users to use AzureClusterIdentites for their v1alpha4 clusters? Should the default template example have it? Related to #1445

- document it in the quick start guide
- add it to the example templates?

**Comment by CecileRobertMichon** [comment]
> 1) document it in the quick start guide

might not be as easy as setting env vars for quick start but if we do 2) the quickstart experience should stay the same

> 2) add it to the example templates?

Yeah I think we should consider doing this

> 3) add some deprecation warning (not sure how)

this we can do in the controller logs and/or as a k8s event whenever credentials are set through manager environment variables

**Comment by nader-ziada** [comment]
I'll just try again before I debug too much 

/test pull-cluster-api-provider-azure-e2e

**Comment by nader-ziada** [comment]
the private cluster test is failing again because its getting stuck on creating the bastion host, 
```
I0622 21:57:02.456316       1 azurebastion.go:46] controllers/AzureCluster "msg"="creating bastion host" "azureCluster"="capz-e2e-jp3qx9" "cluster"="capz-e2e-jp3qx9" "namespace"="capz-e2e-tq69po" "bastion"="capz-e2e-jp3qx9-azure-bastion"
```
it eventually times out the private cluster is never created, 

still looking into it

**Comment by CecileRobertMichon** [comment]
/lgtm

@nader-ziada can you follow up on 

> document it in the quick start guide
might not be as easy as setting env vars for quick start but if we do 2) the quickstart experience should stay the same
add it to the example templates?
Yeah I think we should consider doing this
add some deprecation warning (not sure how)
this we can do in the controller logs and/or as a k8s event whenever credentials are set through manager environment variables

**Comment by nader-ziada** [comment]
> might not be as easy as setting env vars for quick start but if we do 2) the quickstart experience should stay the same



> document it in the quick start guide
> might not be as easy as setting env vars for quick start but if we do 2) the quickstart experience should stay the same
> add it to the example templates?
> Yeah I think we should consider doing this
> add some deprecation warning (not sure how)
> this we can do in the controller logs and/or as a k8s event whenever credentials are set through manager environment variables

@CecileRobertMichon it was thinking to wait until we also can do that for the private clusters so we don't have any exceptions in the quick start guide, wdyt?

**Comment by CecileRobertMichon** [comment]
/approve
/hold 
feel free to unhold whenever we have confirmation that the conformance tests above are creating clusters without issue

**Comment by nader-ziada** [comment]
/hold cancel

@CecileRobertMichon PTAL

all the tests are using AzureClusterIdentity except for `SelfHosted` and `PrivateCluster`

**Comment by nader-ziada** [comment]
rebased on capi v0.4.0

File: azure/scope/identity.go [hunk 2]
```diff
@@ -182,9 +181,9 @@ func (p *AzureCredentialsProvider) GetAuthorizer(ctx context.Context, resourceMa
 			Name:      fmt.Sprintf("%s-binding", copiedIdentity.Name),
 			Namespace: copiedIdentity.Namespace,
 			Labels: map[string]string{
-				clusterv1.ClusterLabelName:         clusterMeta.Name,
-				infrav1.ClusterLabelNamespace:      clusterMeta.Namespace,
-				clusterctl.ClusterctlMoveLabelName: "true",
+				clusterv1.ClusterLabelName:                   clusterMeta.Name,
+				infrav1.ClusterLabelNamespace:                clusterMeta.Namespace,
+				"clusterctl.cluster.x-k8s.io/move-hierarchy": "true",
```
Comment (CecileRobertMichon) [inline azure/scope/identity.go:186]: why more away from using a constant here?

File: test/e2e/azure_test.go [hunk 3]
```diff
@@ -145,6 +168,43 @@ var _ = Describe("Workload cluster creation", func() {
 				})
 			})
 		})
+
+		Context("Running the self-hosted spec", func() {
```
Comment (CecileRobertMichon) [inline test/e2e/azure_test.go:172]: why not keep this in the capi_e2e suite like it was before? It's not really testing functionality that can be regressed by CAPZ PRs

Comment (nader-ziada) [reply test/e2e/azure_test.go:172]: because for the identity secret to be moved when the cluster pivots to a management cluster running on Azure, it has to be in the same namespace as the cluster and running the test spec from capi doesn't allow me to know the namespace

File: azure/scope/identity.go (hunk 2 above)
Comment (nader-ziada) [reply azure/scope/identity.go]: I'll fix that

File: test/e2e/azure_test.go (hunk 3 above)
Comment (CecileRobertMichon) [reply test/e2e/azure_test.go:172]: right but could we define our own spec and run it as part of pull-cluster-api-provider-azure-capi-e2e ?

Comment (nader-ziada) [reply test/e2e/azure_test.go:172]: oh okay, just move the calling it in capi_e2e?

Comment (CecileRobertMichon) [reply test/e2e/azure_test.go:172]: yeah that's what I meant, sorry if it wasn't clear

Comment (nader-ziada) [reply test/e2e/azure_test.go]: no worries, that makes better sense, will change it

**Comment by CecileRobertMichon** [comment]
will this affect scripts/ci-entrypoint.sh? want to make sure we don't break https://testgrid.k8s.io/provider-azure-master-signal jobs that rely on it

**Comment by nader-ziada** [comment]
> will this affect scripts/ci-entrypoint.sh? want to make sure we don't break https://testgrid.k8s.io/provider-azure-master-signal jobs that rely on it

you are right, `hack/create-dev-cluster.sh` doesn't create the secret or set the related `AzureClusterIdentity` fields, so would probably fail

**Comment by nader-ziada** [comment]
@CecileRobertMichon I think the last two commits cover everything,   waiting for tests now🤞

File: hack/create-identity-secret.sh [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+source "${REPO_ROOT}/hack/ensure-kubectl.sh"
+# shellcheck source=hack/parse-prow-creds.sh
+source "${REPO_ROOT}/hack/parse-prow-creds.sh"
+
+export CLUSTER_IDENTITY_SECRET_NAME="cluster-identity-secret"
+export CLUSTER_IDENTITY_NAME=${CLUSTER_IDENTITY_NAME:="cluster-identity"} 
+export AZURE_CLUSTER_IDENTITY_SECRET_NAMESPACE="default"
+export AZURE_CLUSTER_IDENTITY_CLIENT_ID="${AZURE_CLIENT_ID}"
+
+kubectl create secret generic "${CLUSTER_IDENTITY_SECRET_NAME}" --from-literal=clientSecret="${AZURE_CLIENT_SECRET}"
```
Comment (CecileRobertMichon) [inline hack/create-identity-secret.sh:33]: how do we make it obvious that users need to do this on their own? Should we document the steps in getting-started.md and the quickstart?

Comment (nader-ziada) [reply hack/create-identity-secret.sh:33]: I updated the docs in this PR #1477 which refers to this doc here: https://capz.sigs.k8s.io/topics/multitenancy.html

we would need to update the quick start but that's in capi, right?

**Comment by CecileRobertMichon** [comment]
@nader-ziada did you test Tilt with these changes? I believe we also need to update the Tiltfile to create the secret on the management cluster

**Comment by nader-ziada** [comment]
> @nader-ziada did you test Tilt with these changes? I believe we also need to update the Tiltfile to create the secret on the management cluster

I did not try tilt, will try it now

File: templates/test/ci/cluster-template-prow-multi-tenancy.yaml [hunk 5]
```diff
@@ -418,6 +418,18 @@ data:
                             type: string
                         type: object
                       type: array
+                    serviceLoadBalancerIPs:
```
Comment (CecileRobertMichon) [inline templates/test/ci/cluster-template-prow-multi-tenancy.yaml:421]: can we completely get rid of the multi-tenancy flavor now that all flavors are using multi-tenancy?

Comment (nader-ziada) [reply templates/test/ci/cluster-template-prow-multi-tenancy.yaml:421]: I removed it from templates and removed the test, missed this one

File: templates/cluster-template.yaml [hunk 6]
```diff
@@ -193,3 +197,17 @@ spec:
...
           name: '{{ ds.meta_data["local_hostname"] }}'
+---
+apiVersion: infrastructure.cluster.x-k8s.io/v1alpha4
+kind: AzureClusterIdentity
+metadata:
+  name: ${CLUSTER_IDENTITY_NAME}
+  namespace: default
+spec:
+  allowedNamespaces: {}
+  clientID: ${AZURE_CLUSTER_IDENTITY_CLIENT_ID}
```
Comment (CecileRobertMichon) [inline templates/cluster-template.yaml:208]: would it make sense to keep this as `AZURE_CLIENT_ID` so existing setups (tilt, etc) can reuse the existing var?

Comment (nader-ziada) [reply templates/cluster-template.yaml:208]: let me try

File: Tiltfile [hunk 7]
```diff
@@ -169,6 +169,15 @@ def capz():
 
     k8s_yaml(blob(yaml))
 
+def create_identity_secret():
+    #create secret for identity password
+    os.putenv('CLUSTER_IDENTITY_SECRET_NAME', 'cluster-identity-secret')
```
Comment (CecileRobertMichon) [inline Tiltfile:174]: will we need to add the namespace variable to tilt-settings (if so needs to be added to tilt docs), or could we just default it / set it here?

Comment (nader-ziada) [reply Tiltfile:174]: its defaulting to `default` here currently, same as the script `hack/create-identity-secret.sh` 

for the e2e tests, its in the same namespace as the cluster

not sure what we need to change?

Comment (CecileRobertMichon) [reply Tiltfile:174]: it looks like the templates expect `AZURE_CLUSTER_IDENTITY_SECRET_NAMESPACE` to be set, should that be set here? Same question for `CLUSTER_IDENTITY_NAME`.

Basically, I'm wondering if I pull the changes from this PR and run `make tilt-up` and then click on one of the flavors to deploy it, will everything work as it did before? or are there any additional vars that need to be defined in my tilt-settings.json file now? It would be nice if everything kept working as before.

Comment (nader-ziada) [reply Tiltfile:174]: I think it should work the same, I added the setting of the namespace and identity name in the tiltfile similar to the script

**Comment by CecileRobertMichon** [comment]
@nader-ziada tried out tilt locally, found two issues:
- if you run tilt up a second time it fails with `error: failed to create secret secrets "cluster-identity-secret" already exists`: we should delete the secret and recreate it every tilt up so it's idempotent (same thing we do with configmaps)
- the secret env var is getting printed in tilt logs, any way we can avoid that?

File: hack/create-identity-secret.sh [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+set -o errexit
+set -o nounset
+set -o pipefail
+
```
Comment (CecileRobertMichon) [inline hack/create-identity-secret.sh:19]: `set +o xtrace`

we want to make sure the secret doesn't get printed

**Comment by CecileRobertMichon** [comment]
found https://github.com/tilt-dev/tilt-extensions/tree/master/secret, not sure if it would be better to use that to create the secret (don't know if it logs the secret value the same way)

File: Tiltfile (hunk 7 above)
Comment (nader-ziada) [reply Tiltfile:174]: > @nader-ziada tried out tilt locally, found two issues:
> 
> * if you run tilt up a second time it fails with `error: failed to create secret secrets "cluster-identity-secret" already exists`: we should delete the secret and recreate it every tilt up so it's idempotent (same thing we do with configmaps)
> * the secret env var is getting printed in tilt logs, any way we can avoid that?

where does it delete configmaps? can't find that

Comment (CecileRobertMichon) [reply Tiltfile:174]: here: https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/master/Tiltfile#L174

**Comment by nader-ziada** [comment]
> @nader-ziada tried out tilt locally, found two issues:
> 
> * if you run tilt up a second time it fails with `error: failed to create secret secrets "cluster-identity-secret" already exists`: we should delete the secret and recreate it every tilt up so it's idempotent (same thing we do with configmaps)
> * the secret env var is getting printed in tilt logs, any way we can avoid that?

- added the delete statement similar to the configmaps
- added the quiet=true option to prevent printing

File: Tiltfile [hunk 9]
```diff
@@ -169,6 +169,20 @@ def capz():
...
+def create_identity_secret():
+    #create secret for identity password
+    local("kubectl delete secret cluster-identity-secret --ignore-not-found=true")
+
+    os.putenv('AZURE_CLUSTER_IDENTITY_SECRET_NAME', 'cluster-identity-secret')
+    os.putenv('AZURE_CLUSTER_IDENTITY_SECRET_NAMESPACE', 'default')
+    os.putenv('CLUSTER_IDENTITY_NAME', 'cluster-identity')
+
+    substitutions = settings.get("kustomize_substitutions", {})
+    os.environ.update(substitutions)
```
Comment (CecileRobertMichon) [inline Tiltfile:181]: what is this update doing?

File: hack/create-identity-secret.sh [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+source "${REPO_ROOT}/hack/ensure-kubectl.sh"
+# shellcheck source=hack/parse-prow-creds.sh
+source "${REPO_ROOT}/hack/parse-prow-creds.sh"
+
+export AZURE_CLUSTER_IDENTITY_SECRET_NAME="cluster-identity-secret"
+export CLUSTER_IDENTITY_NAME=${CLUSTER_IDENTITY_NAME:="cluster-identity"} 
+export AZURE_CLUSTER_IDENTITY_SECRET_NAMESPACE="default"
+
+kubectl create secret generic "${AZURE_CLUSTER_IDENTITY_SECRET_NAME}" --from-literal=clientSecret="${AZURE_CLIENT_SECRET}"
+kubectl label secret "${AZURE_CLUSTER_IDENTITY_SECRET_NAME}" "clusterctl.cluster.x-k8s.io/move-hierarchy=true" --overwrite=true
```
Comment (devigned) [inline hack/create-identity-secret.sh:34]: New line

File: hack/gen-flavors.sh [hunk 11]
```diff
@@ -17,15 +17,15 @@
 set -o errexit
 set -o nounset
 set -o pipefail
-#set -o verbose
+# set -o verbose
```
Comment (devigned) [inline hack/gen-flavors.sh:20]: delete?

File: templates/flavors/aks/kustomization.yaml [hunk 12]
```diff
@@ -3,3 +3,6 @@ kind: Kustomization
 namespace: default
 resources:
 - cluster-template.yaml
+- ../../azure-cluster-identity
+patchesStrategicMerge:
+- ../../azure-cluster-identity/managedazurecluster-identity-ref.yaml
```
Comment (devigned) [inline templates/flavors/aks/kustomization.yaml:8]: newline

File: templates/flavors/system-assigned-identity/kustomization.yaml [hunk 13]
```diff
@@ -1,5 +1,6 @@
 namespace: default
 resources:
-  - ../default
+  - ../base
+  - ../default/machine-deployment.yaml
 patchesStrategicMerge:
   - patches/system-assigned-identity.yaml
```
Comment (devigned) [inline templates/flavors/system-assigned-identity/kustomization.yaml:6]: newline

File: templates/flavors/user-assigned-identity/kustomization.yaml [hunk 14]
```diff
@@ -1,5 +1,6 @@
 namespace: default
 resources:
-  - ../default
+  - ../base
+  - ../default/machine-deployment.yaml
 patchesStrategicMerge:
   - patches/user-assigned-identity.yaml
```
Comment (devigned) [inline templates/flavors/user-assigned-identity/kustomization.yaml:6]: newline

File: templates/flavors/windows/kustomization.yaml [hunk 15]
```diff
@@ -3,4 +3,8 @@ resources:
   - ../base-windows
   - machine-deployment.yaml
   - machine-deployment-windows.yaml
+  - ../../azure-cluster-identity
+
+patchesStrategicMerge:
+  - ../../azure-cluster-identity/azurecluster-identity-ref.yaml
```
Comment (devigned) [inline templates/flavors/windows/kustomization.yaml:10]: weird newline

File: templates/test/ci/prow-ci-version/kustomization.yaml [hunk 16]
```diff
@@ -7,3 +7,12 @@ patchesStrategicMerge:
...
   - patches/machine-deployment-ci-version.yaml
+apiVersion: kustomize.config.k8s.io/v1beta1
+kind: Kustomization
+namespace: default
+resources:
+  - ../prow
+patchesStrategicMerge:
+  - ../patches/control-plane-ci-version.yaml
+  - ../patches/controller-manager.yaml
+  - patches/machine-deployment-ci-version.yaml
```
Comment (devigned) [inline templates/test/ci/prow-ci-version/kustomization.yaml:18]: newline

File: templates/test/ci/prow-identity-from-env/mhc.yaml [hunk 17]
```diff
@@ -0,0 +1,15 @@
...
+spec:
+  clusterName: "${CLUSTER_NAME}"
+  maxUnhealthy: 100%
+  selector:
+    matchLabels:
+      nodepool: "pool1"
+  unhealthyConditions:
+    - type: E2ENodeUnhealthy
+      status: "True"
+      timeout: 30s
```
Comment (devigned) [inline templates/test/ci/prow-identity-from-env/mhc.yaml:15]: newline

**Review by devigned — CHANGES_REQUESTED** [review]
All nits.

**Comment by CecileRobertMichon** [comment]
Do we want to keep all 7 commits in the final merge or are some of these squashable?

**Comment by nader-ziada** [comment]
> Do we want to keep all 7 commits in the final merge or are some of these squashable?

I think they are for different parts, but I can squash if you prefer

**Comment by CecileRobertMichon** [comment]
> I think they are for different parts, but I can squash if you prefer

nope, all good, just checking they are distinct

**Comment by nader-ziada** [comment]
applied @devigned comments about newlines

**Review by devigned — APPROVED** [review]
/lgtm

**Comment by nader-ziada** [comment]
thanks for all the help with debugging this PR @CecileRobertMichon @devigned @mboersma

**Comment by devigned** [comment]
@nader-ziada great work on this one! This one was a tough task. Thank you for taking this on.
